File: Arrays/Algorithms/Sliding_Window/Two_Pointer/longest_substring_without_repeating_chars.cpp
#include <bits/stdc++.h>
using namespace std;

int lengthOfLongestSubstring(const string& s) {
	unordered_map<char, int> last;
	int l = 0, res = 0;
	for (int r = 0; r < s.size(); r++) {
		if (last.count(s[r])) l = max(l, last[s[r]] + 1);
		last[s[r]] = r;
		res = max(res, r - l + 1);
	}
	return res;
}

int main() {
	string s;
	cin >> s;
	cout << lengthOfLongestSubstring(s) << endl;
	return 0;
}

File: Arrays/Algorithms/Sliding_Window/Two_Pointer/longest_substring_without_repeating_chars.py
def length_of_longest_substring(s):
	last = {}
	l = res = 0
	for r, c in enumerate(s):
		if c in last:
			l = max(l, last[c] + 1)
		last[c] = r
		res = max(res, r - l + 1)
	return res

if __name__ == "__main__":
	s = input().strip()
	print(length_of_longest_substring(s))
